clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -502,8 +502,6 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
 #undef VISITCOMP
 
   mlir::Value VisitBinAssign(const BinaryOperator *E);
-  mlir::Value VisitBinLAnd(const BinaryOperator *E) { llvm_unreachable("NYI"); }
```

Comment:
As a first principle for doing anything with ClangIR we start from `do what codegen does`. It's pretty much the only way (in our opinion) to be able to guarantee that we don't end up materially change the behavior of clang.

For the example here, clang's `ScalarExprEmitter::VisitBinLAnd` checks if it's looking at a vector type:
https://github.com/llvm/clangir/blob/06022c5a738db8aa967c8dd89e82b8a2a9df8fea/clang/lib/CodeGen/CGExprScalar.cpp#L4525

We don't yet care about vector types, so we'd just assert that we're not seeing this via something like this:

```
Value *ScalarExprEmitter::VisitBinLAnd(const BinaryOperator *E) {
  // Perform vector logical and on comparisons with zero vectors.
  if (E->getType()->isVectorType()) {
    llvm_unreachable("Vector types aren't yet supported in ClangIR");
  }
```

And then where it checks if we're enabling instr profiling you could create a method in https://github.com/llvm/clangir/blob/main/clang/lib/CIR/CodeGen/UnimplementedFeatureGuarding.h and assert that it's false:

```
  if (CGF.CGM.getCodeGenOpts().hasProfileClangInstr() && UnimplementedFeature::hasProfileClangInstr())
    llvm_unreachable("ProfileClangInstr NYI");
```

The approach to having lor and land in CIR seems reasonable to me. TBH I don't know why those ops don't exist in LLVM. @bcardosolopes since he's been around longer than me and might know.

Reply:
As to the design of this however I'm not sure a simple cir.binop will work. C++ requires that the operation short circuits

```
if (a() && b())
  ...
```

If `a()` is true then `b()` never happens. Your current approach would cirgen `b()` prior to the binop and thus would be evaluated.  

We've been loosely thinking about some sort of

```
%land.result = cir.logical and { // lhs
  %a = cir.call a
} rhs {
  %b = cir.call b
}
```

for this. Maybe even just using `cir.if`.

Reply:
Thanks for the insight. I only thought of short-circuiting when lowering CIR to LLVM and didn't thought of CIR codegen at all. I think you brought up with a very good example with function call return values used as logical operator operands.  My current approach doesn't produce correct result if those functions have side-effects.

I like the idea of using regions to represent operands and it allows the construction of `a && ((b + 1) == 3)`, where the operand can be arbitrary expressions. One thing I would suggest to add is using `cir.yield` as the region branch terminator.

If you both of you think this is a good idea, I can start working on it.